SD-1297 - Improvements to Spree::Menu

MatthewKennedy opened the pull request:
## Menu Locale Fallback

Add better fallback behaviour for menus when viewer is visiting the site in a language that has not had a menu created for the location and language of the store is set to.

- Menu falls back to `current_store.default_locale`
- Add admin view indicator to show default menus and prompt users to create default menus.
- Add tests for this behaviour

## Make Links Reusable Outside of Menu Items
As work on CMS Pages continues it is clear that linking to products and taxons will be done from more than just the navigation menu items. For example hero sections containing clickable images or buttons.

- Move Link method from `Spree::MenuItem` to a concern `Spree::DisplayLink`.
- Move view file link selector to shared views for use in various parts of the admin.
- Rename a frontend helper method so it does not sound so menu item specific.
- Fix admin javascript for link switcher so that it is not tied to use in menu item linking.

I have made these changes on this PR so that if these tests pass and it is all good here. The CMS Pages PR is then more focused on its own new features and is not changing existing behaviour of the menu whilst adding the whole new feature.

## Optional Pop-up Information
Add in ability to trigger an info pop-up alert that is toggled by clicking an info icon, the info message stays on screen until the close button is clicked giving the viewer lots of time to read the message.

These info pop-up alerts might not be perfect in all scenarios, but they are good when you want a message that might only need reading once, and not need displaying in the DOM all the time.

<img width="1198" alt="Screenshot 2021-05-23 at 12 04 01" src="https://user-images.githubusercontent.com/1240766/119257847-175c3a80-bbbf-11eb-82bb-5d410c2a0f5d.png">

MatthewKennedy committed:
Move link method from Spree::MenuItem to a concern.

MatthewKennedy committed:
Add menu fallback to current_store.default_locale

MatthewKennedy committed:
Add visual indication of default menu for store.

MatthewKennedy committed:
Add helpful hint pop-up for default menus.

MatthewKennedy committed:
Update naming of link.rb to display_link.rb

MatthewKennedy committed:
Improve containers info message

MatthewKennedy committed:
Improve view setup for link types.

MatthewKennedy committed:
Refactor links to be independent of menu items for use in other parts of CMS.

MatthewKennedy committed:
Improve JavaScript wording.

MatthewKennedy committed:
Look for correct translation in tests.

damianlegawiec commented on backend/app/helpers/spree/admin/menu_helper.rb:
```diff
@@ -47,6 +47,14 @@ def menu_locations_dropdown_values
         menu_items_for_select
       end
+
+      def default_menu_for_store?(menu)
```
@MatthewKennedy can we just have `default_menu(location)` method on the `Store` model?

MatthewKennedy replied:
I'll implement that now.

MatthewKennedy replied:
Hi @damianlegawiec 

The highlighted method in the comment above is just a simple view helper method for the admin UI that returns true of false to show if the menu is default in the admin UI Index page.

I added the `default_menu(location)` and used that in the fallback code for the menu, if you want to take a look at the changes?

mad-eel commented on backend/app/assets/javascripts/spree/backend/cms/links.es6:
```diff
@@ -0,0 +1,24 @@
+document.addEventListener('DOMContentLoaded', function() {
+  const linkSwitcher = $('.link_switcher').select2()
```
AFAIK we shouldn't be using jQuery anymore as we want to stop using it in spree 5 entirely.
@damianlegawiec am I correct?

damianlegawiec replied:
@mad-eel we should avoid, however unfortunately `select2` requires jquery - we would need to think how to resolve in the future (change the library, check select2 maintainers plans for future development and dropping jquery dependency and so on)

MatthewKennedy replied:
Hi @mad-eel 

Select2 is a jQuery based lib, if we are using select2 when it move to V5 then I can re-write all this in plane JS

MatthewKennedy committed:
Add default_menu(location) method to Store model.

damianlegawiec commented on core/app/models/spree/store.rb:
```diff
@@ -58,6 +58,10 @@ def self.available_locales
     end
 
+    def default_menu(location)
+      menus.find_by(location: location, locale: default_locale)
```
```suggestion
      menus.find_by(location: location, locale: default_locale) || menus.find_by(location: location)
```

so if there's no menu with this locale fallback the the first one?

MatthewKennedy replied:
Now returns the first created menu for the location as a last, last resort: `find_by(location: location_name, created_at: :asc)`

MatthewKennedy replied:
`default_scope { order(created_at: :asc) }` seems to work.

mad-eel replied:
@MatthewKennedy I also think it's work moving this line here: https://github.com/spree/spree/pull/11051/files#diff-28063c4275ea907ab47a79d37844ab42ca50fe09f376f958745615b3b8d45363R36

It seems to me this belongs to being the "default menu". WDYT?

MatthewKennedy replied:
Hi @mad-eel 

Yeah I'll move that over and test it out, it is a busy section of code that.

damianlegawiec commented on frontend/app/helpers/spree/navigation_helper.rb:
```diff
@@ -57,7 +57,7 @@ def spree_menu(location = 'header')
 
       if available_menus.respond_to?(method_name) && Spree::Menu::MENU_LOCATIONS_PARAMETERIZED.include?(location)
-        available_menus.send(method_name, I18n.locale, current_store.default_locale)
+        available_menus.send(method_name, current_store)
```
@MatthewKennedy we still should pass the locale here so if we have multi-language store we should show the menu with this locale or fallback to the default one

MatthewKennedy replied:
@damianlegawiec I moved `I18n.locale` to the receiving end of this method in the Menu model, is that not advised?

MatthewKennedy commented on core/app/models/spree/menu.rb:
```diff
@@ -27,9 +27,11 @@ class Menu < Spree::Base
-        menu = find_by(location: name, locale: locale.to_s) || find_by(location: name)
+    MENU_LOCATIONS_PARAMETERIZED.each do |location_name|
+      define_singleton_method("for_#{location_name}") do |store|
+        menu = find_by(location: location_name, locale: I18n.locale.to_s) ||
```
@damianlegawiec Is it best to keep `I18n.locale` passed in from the frontend?

MatthewKennedy committed:
Move I18n.local to front end.
Set menu default_scope to return descending order on the created_at for the oldest menu edge case fallback.

MatthewKennedy committed:
Switch from :desc to :asc

MatthewKennedy committed:
Set :asc only on the fallback call.

MatthewKennedy committed:
Revert change.

MatthewKennedy committed:
Move menu last resort fallback to Spree::Store.default_menu

damianlegawiec commented on core/app/models/concerns/spree/display_link.rb:
```diff
@@ -0,0 +1,28 @@
+      when 'Spree::Product'
+        return if linked_resource.nil?
+
+        if defined?(SpreeFrontend)
```
```suggestion
        if Spree::Core::Engine.routes.url_helpers.method_defined?(: product_path)
```

MatthewKennedy replied:
On it...

damianlegawiec commented on core/app/models/concerns/spree/display_link.rb:
```diff
@@ -0,0 +1,28 @@
+      when 'Spree::Taxon'
+        return if linked_resource.nil?
+
+        if defined?(SpreeFrontend)
```
@MatthewKennedy same here

MatthewKennedy committed:
Check for actual routes being set, not just the presence of SpreeFrontend.

damianlegawiec commented on frontend/app/helpers/spree/navigation_helper.rb:
```diff
@@ -57,11 +57,11 @@ def spree_menu(location = 'header')
 
       if available_menus.respond_to?(method_name) && Spree::Menu::MENU_LOCATIONS_PARAMETERIZED.include?(location)
-        available_menus.send(method_name, I18n.locale)
+        available_menus.send(method_name, I18n.locale, current_store)
```
@MatthewKennedy why do we need to pass `current_store` here? `available_menus` is already scoped to the current store: https://github.com/spree/spree/blob/5ecb98271d927050cc08a0aafda4593a874ce841/core/lib/spree/core/controller_helpers/store.rb#L17

MatthewKennedy replied:
I was using that for calling `store.default_menu(location_name)` inside `Spree::Menu`, for the fallback.

MatthewKennedy replied:
@damianlegawiec How about the latest commit as an option?

MatthewKennedy committed:
Find default menu

MatthewKennedy committed:
Make helper_method available.

MatthewKennedy committed:
Update method name.

MatthewKennedy committed:
Use spree_nav_link_tag in mobile menu.

MatthewKennedy committed:
Add notes for code and image.

MatthewKennedy committed:
Dedicated info pop up message.

MatthewKennedy committed:
Use current_store.default_menu(location)

Vegann commented on core/spec/models/spree/menu_spec.rb:
```diff
@@ -31,6 +31,30 @@
+    let!(:menu_en) { create(:menu, name: 'Main Menu EN', store: fr_store, locale: 'en') }
+    let!(:menu_fr) { create(:menu, name: 'Main Menu FR', store: fr_store, locale: 'fr') }
+
+    context 'when the I18n.locale is set to a none default laguage and a menu is available' do
```
```suggestion
    context 'when the I18n.locale is set to a none default language and a menu is available' do
```

Vegann commented on core/spec/models/spree/menu_spec.rb:
```diff
@@ -31,6 +31,30 @@
+      end
+    end
+
+    context 'when the I18n.locale is set to a none default laguage and a menu is not available' do
```
```suggestion
    context 'when the I18n.locale is set to a none default language and a menu is not available' do
```

Vegann commented on core/spec/models/spree/menu_spec.rb:
```diff
@@ -31,6 +31,30 @@
+      end
+    end
+
+    context 'when the I18n.locale is set to the default laguage and a menu is available' do
```
```suggestion
    context 'when the I18n.locale is set to the default language and a menu is available' do
```

MatthewKennedy committed:
Fix spelling in spec tests.

MatthewKennedy commented on core/app/models/concerns/spree/display_link.rb:
```diff
@@ -0,0 +1,28 @@
+      when 'Spree::Taxon'
+        return if linked_resource.nil?
+
+        if Spree::Core::Engine.routes.url_helpers.method_defined?(:nested_taxons_path)
+          Spree::Core::Engine.routes.url_helpers.nested_taxons_path(linked_resource.permalink)
+        else
+          "/#{Spree::Config[:storefront_taxons_path]}/#{linked_resource.permalink}"
+        end
+      when 'Spree::Product'
+        return if linked_resource.nil?
+
+        if Spree::Core::Engine.routes.url_helpers.method_defined?(:products_path)
+          Spree::Core::Engine.routes.url_helpers.product_path(linked_resource)
+        else
+          "/#{Spree::Config[:storefront_products_path]}/#{linked_resource.slug}"
+        end
+      when 'Home Page'
+        '/'
+      when 'URL'
+        destination
+      end
```
I'll get that sorted!

damianlegawiec commented on core/app/models/spree/store.rb:
```diff
@@ -58,6 +58,14 @@ def self.available_locales
+    def default_menu(location)
+      menu = menus.find_by(location: location, locale: default_locale) || menus.find_by(location: location)
+
+      if menu.present?
+        menu.root
+      end
```
```suggestion
      menu.root if f menu.present?
```

damianlegawiec commented on backend/app/views/spree/admin/shared/linkables/_spree_product.html.erb:
```diff
@@ -0,0 +1,23 @@
+<% end %>
+
+<script>
+  document.addEventListener('DOMContentLoaded', function() {
```
@MatthewKennedy can we have one autocomplete setup per app and get the values for `data_attrbute_name`, `apiUrl` etc from `data-` HTML tags? this will help us later move gradually to Stimulus and Hotwire :)

MatthewKennedy replied:
Yeah that works for me, I'll sort that, I prefer that approach myself.

Do you want that change on a fresh PR?

MatthewKennedy replied:
Actually no, I'll do it on this PR then I can implement it straight onto the cms_pages if that's okay?

MatthewKennedy committed:
Use private method for spree_routes

MatthewKennedy committed:
Set up JavaScript adapter for use with HTML data attributes.

damianlegawiec commented on backend/app/views/spree/admin/shared/linkables/_spree_taxon.html.erb:
```diff
@@ -6,19 +6,17 @@
+                                           id_holder, name_holder, active_id_holder || nil),
+        { include_blank: true },
+        class: 'autocomplete',
+        data: { ac_name: 'taxons',
+                ac_placeholder: Spree.t('admin.navigation.seach_for_a_taxon'),
+                ac_clear: true,
+                ac_url: 'taxons_api_v2',
+                ac_return_attr: 'pretty_name' } %>
```
```suggestion
        data: { 'autocomplete-name-value': 'taxons',
                'autocomplete-placeholder-value': Spree.t('admin.navigation.seach_for_a_taxon'),
                'autocomplete-clear-value': true,
                'autocomplete-url-value': 'taxons_api_v2',
                'autocomplete-return-attribute-value': 'pretty_name' } %>
```

damianlegawiec replied:
that's the Stimulus conventions :) 

https://stimulus.hotwire.dev/reference/values

this way we would be able to switch to stimulus quote quickly

MatthewKennedy replied:
Hi @damianlegawiec,

The HTML output is kebab-case: `data-ac-name="products" `

```html
<select class="select2autocomplete select2-hidden-accessible" data-ac-name="products" data-ac-placeholder="Search for a Product" data-ac-clear="true" data-ac-url="products_api_v2" name="menu_item[linked_resource_id]" id="menu_item_linked_resource_id" data-select2-id="menu_item_linked_resource_id" tabindex="-1" aria-hidden="true"><option value="" label=" " data-select2-id="2"></option>
<option selected="selected" value="17" data-select2-id="3">V Neck T Shirt</option></select>
```

I'll update this now to use the full name `autocomplete` and end the in `-value` and then I think it should be fully compatible.

MatthewKennedy replied:
```html
<select class="select2autocomplete select2-hidden-accessible" data-autocomplete-name-value="taxons" data-autocomplete-placeholder-value="Search for a Taxon" data-autocomplete-clear-value="true" data-autocomplete-url-value="taxons_api_v2" data-autocomplete-return-attr-value="pretty_name" name="menu_item[linked_resource_id]" id="menu_item_linked_resource_id" data-select2-id="menu_item_linked_resource_id" tabindex="-1" aria-hidden="true"><option value="" label=" " data-select2-id="2"></option>
<option selected="selected" value="21" data-select2-id="3">Categories -&gt; Streetstyle</option></select>
```

MatthewKennedy committed:
Avoid name clash with existing auto complete class.

MatthewKennedy committed:
Fix commenting in JS file.

MatthewKennedy committed:
Update return clause for menu.root

MatthewKennedy committed:
Update naming for Stimulus compatibility.
